src/virtual_file_store.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 The Amber Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_VIRTUAL_FILE_STORE_H_
+#define SRC_VIRTUAL_FILE_STORE_H_
+
+#include "amber/result.h"
+
+#include <string>
+#include <unordered_map>
+
+namespace amber {
+
+/// Stores a number of virtual files by path.
+class VirtualFileStore {
+ public:
+  /// Return the path sanitized into a canonical form.
+  static std::string GetCanonical(const std::string& path);
+
+  /// Adds the virtual file with content |content| to the virtual file path
+  /// |path|. If there's already a virtual file with the given path, an error is
+  /// returned.
+  Result Add(const std::string& path, const std::string& content) {
+    auto canonical = GetCanonical(path);
+
+    auto it = files_by_path_.find(canonical);
+    if (it != files_by_path_.end()) {
+      return Result("Virtual file '" + path + "' already declared");
+    }
+    files_by_path_.emplace(canonical, content);
+    return {};
+  }
+
+  /// Look up the virtual file by path. If the file was found, the content is
+  /// assigned to content.
+  Result Get(const std::string& path, std::string* content) const {
+    auto canonical = GetCanonical(path);
```

Comment:
Should we either `assert(content)` or check and return failure if it's nullptr?

Reply:
Do you mean zero length? I could certainly do this in `Set`. I think the `Get` handler in this case is probably fine (it won't be found).

Reply:
Content is a pointer, so what happens if I call `Get("foo", nullptr)` ?

Reply:
Done on both.

Reply:
Oh you mean `content`. Yeah, could `assert()` - not sure it'll be _that much_ clearer than the null-pointer segfault you'll get. :)

If this is a standard thing done in amber for out variables, I can add.

Reply:
I saw some other places that does indeed `assert()` on out parameters. Done.

Reply:
I don't remember what the standard is in amber, but `assert(content && "content should not be nullptr")` can be clearer, heh

Reply:
I've seen some compilers warn about `&&`'ing with string literals, so I'll opt to keep as a basic `assert(content)` if that's okay.

Reply:
Works for me.